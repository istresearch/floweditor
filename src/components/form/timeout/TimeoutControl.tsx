import { react as bindCallbacks } from 'auto-bind';
import CheckboxElement from 'components/form/checkbox/CheckboxElement';
import * as React from 'react';
import { renderIf } from 'utils';

import styles from './TimeoutControl.module.scss';
import i18n from 'config/i18n';
import TembaSelect, { TembaSelectStyle } from 'temba/TembaSelect';
import { SelectOption } from '../select/SelectElement';

export const TIMEOUT_OPTIONS: SelectOption[] = [
  { value: '60', name: i18n.t('forms.timeout_1 minute', '1 minute') },
  { value: '120', name: i18n.t('forms.timeout_2 minutes', '2 minutes') },
  { value: '180', name: i18n.t('forms.timeout_3 minutes', '3 minutes') },
  { value: '240', name: i18n.t('forms.timeout_4 minutes', '4 minutes') },
  { value: '300', name: i18n.t('forms.timeout_5 minutes', '5 minutes') },
  { value: '600', name: i18n.t('forms.timeout_10 minutes', '10 minutes') },
  { value: '900', name: i18n.t('forms.timeout_15 minutes', '15 minutes') },
  { value: '3600', name: i18n.t('forms.timeout_1 hour', '1 hour') },
  { value: '7200', name: i18n.t('forms.timeout_2 hours', '2 hours') },
  { value: '10800', name: i18n.t('forms.timeout_3 hours', '3 hours') },
  { value: '21600', name: i18n.t('forms.timeout_6 hours', '6 hours') },
  { value: '43200', name: i18n.t('forms.timeout_12 hours', '12 hours') },
  { value: '64800', name: i18n.t('forms.timeout_18 hours', '18 hours') },
  { value: '86400', name: i18n.t('forms.timeout_1 day', '1 day') },
  { value: '172800', name: i18n.t('forms.timeout_2 days', '2 days') },
  { value: '259200', name: i18n.t('forms.timeout_3 days', '3 days') },
  { value: '604800', name: i18n.t('forms.timeout_1 week', '1 week') }
];

export const DEFAULT_TIMEOUT = TIMEOUT_OPTIONS[4];

export const ellipsize = (str: string) => `${str}...`;

export interface TimeoutControlProps {
  timeout: number;
  onChanged(timeout: number): void;
}

export default class TimeoutControl extends React.Component<TimeoutControlProps> {
  constructor(props: TimeoutControlProps) {
    super(props);
    bindCallbacks(this, {
      include: [/^handle/]
    });
  }

  private getSelected(timeout: number): any {
    for (const [idx, { value }] of TIMEOUT_OPTIONS.entries()) {
      if (value === '' + timeout) {
        return TIMEOUT_OPTIONS[idx];
      }
    }
    return null;
  }

  private isChecked(): boolean {
    return this.props.timeout > 0;
  }

  private getInstructions(): string {
    const base = i18n.t('forms.continue_when_no_response', 'Continue when there is no response');
    return this.isChecked() ? `${base} for` : ellipsize(base);
  }

  private handleChecked(): void {
    if (this.props.timeout > 0) {
      this.props.onChanged(0);
    } else {
      this.props.onChanged(parseInt(DEFAULT_TIMEOUT.value));
    }
  }

  private handleTimeoutChanged(selected: any): void {
    this.props.onChanged(parseInt(selected.value));
  }

  public render(): JSX.Element {
    return (
      <div className={styles.timeout_control_container}>
        <div className={styles.left_section}>
          <CheckboxElement
            name={i18n.t('forms.timeout', 'Timeout')}
            checked={this.isChecked()}
            description={this.getInstructions()}
            checkboxClassName={styles.checkbox}
            onChange={this.handleChecked}
          />
        </div>
        {renderIf(this.isChecked())(
          <div className={styles.drop_down}>
            <TembaSelect
              name={i18n.t('forms.timeout', 'Timeout')}
              style={TembaSelectStyle.small}
              value={this.getSelected(this.props.timeout)}
              options={TIMEOUT_OPTIONS}
              onChange={this.handleTimeoutChanged}
            ></TembaSelect>
          </div>
        )}
      </div>
    );
  }
}
