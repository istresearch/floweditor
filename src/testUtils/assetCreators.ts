import { languageToAsset } from 'components/flow/actions/updatecontact/helpers';
import { determineTypeConfig } from 'components/flow/helpers';
import { ActionFormProps, LocalizationFormProps, RouterFormProps } from 'components/flow/props';
import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import { DefaultExitNames } from 'components/flow/routers/constants';
import { ResolvedRoutes, resolveRoutes } from 'components/flow/routers/helpers';
import { Methods } from 'components/flow/routers/webhook/helpers';
import { DEFAULT_OPERAND, GROUPS_OPERAND, SCHEMES_OPERAND } from 'components/nodeeditor/constants';
import { Operators, Types } from 'config/interfaces';
import { getTypeConfig, Scheme } from 'config/typeConfigs';
import {
  AnyAction,
  BroadcastMsg,
  CallResthook,
  CallWebhook,
  Case,
  Category,
  ChangeGroups,
  Contact,
  Exit,
  Field,
  Flow,
  FlowNode,
  Group,
  Label,
  OpenTicket,
  PlayAudio,
  RemoveFromGroups,
  Router,
  RouterTypes,
  SayMsg,
  SendEmail,
  SendMsg,
  SetContactChannel,
  SetContactField,
  SetContactLanguage,
  SetContactProperty,
  SetRunResult,
  StartFlow,
  StartFlowArgs,
  StartFlowExitNames,
  StartSession,
  SwitchRouter,
  TransferAirtime,
  UINode,
  Wait,
  WaitTypes,
  WebhookExitNames,
  HintTypes
} from 'flowTypes';
import Localization from 'services/Localization';
import { Asset, Assets, AssetType, RenderNode } from 'store/flowContext';
import { assetListToMap } from 'store/helpers';
import { EMPTY_TEST_ASSETS } from 'test/utils';
import { mock } from 'testUtils';
import * as utils from 'utils';

const { results: groupsResults } = require('test/assets/groups.json');
const languagesResults = require('test/assets/languages.json');
mock(utils, 'createUUID', utils.seededUUIDs());
/**
 * Create a select control option
 */
export const createSelectOption = ({ label }: { label: string }) => ({
  label: utils.capitalize(label.trim()),
  labelKey: 'name',
  valueKey: 'id'
});

export const createSayMsgAction = ({
  uuid = utils.createUUID(),
  text = 'Welcome to Moviefone!'
}: {
  uuid?: string;
  text?: string;
} = {}): SayMsg => ({
  type: Types.say_msg,
  uuid,
  text
});

export const createPlayAudioAction = ({
  uuid = utils.createUUID(),
  audio_url = '/my_audio.mp3'
}: {
  uuid?: string;
  text?: string;
  audio_url?: string;
} = {}): PlayAudio => ({
  type: Types.play_audio,
  uuid,
  audio_url
});

export const createSendMsgAction = ({
  uuid = utils.createUUID(),
  text = 'Hey!',
  all_urns = false,
  labels = []
}: {
  uuid?: string;
  text?: string;
  // tslint:disable-next-line:variable-name
  all_urns?: boolean;
  labels?: Label[];
} = {}): SendMsg => ({
  type: Types.send_msg,
  uuid,
  text,
  all_urns,
  labels
});

export const createSendEmailAction = ({
  uuid = utils.createUUID(),
  subject = 'New Sign Up',
  body = '@run.results.name just signed up.',
  addresses = ['jane@example.com']
}: {
  uuid?: string;
  subject?: string;
  body?: string;
  addresses?: string[];
} = {}): SendEmail => ({
  uuid,
  type: Types.send_email,
  subject,
  body,
  addresses
});

export const createTransferAirtimeAction = ({
  uuid = utils.createUUID()
}: {
  uuid?: string;
} = {}): TransferAirtime => ({
  uuid,
  type: Types.transfer_airtime,
  amounts: {
    USD: 1.5
  },
  result_name: 'Result'
});

export const createCallResthookAction = ({
  uuid = utils.createUUID(),
  resthook = 'my-resthook',
  result_name = 'result'
}: {
  uuid?: string;
  resthook?: string;
  result_name?: string;
} = {}): CallResthook => ({
  uuid,
  type: Types.call_resthook,
  resthook,
  result_name
});

export const createCallWebhookAction = ({
  uuid = utils.createUUID(),
  url = 'https://www.example.com',
  method = Methods.GET,
  result_name = 'result_name'
}: {
  uuid?: string;
  url?: string;
  method?: Methods;
  result_name?: string;
} = {}): CallWebhook => ({
  uuid,
  type: Types.call_webhook,
  url,
  method,
  result_name
});

export const createStartSessionAction = ({
  uuid = utils.createUUID(),
  groups = [{ uuid: utils.createUUID(), name: 'Cat Fanciers' }],
  contacts = [
    { uuid: utils.createUUID(), name: 'Norbert Kwizera' },
    { uuid: utils.createUUID(), name: 'Rowan Seymour' }
  ],
  flow = {
    uuid: 'flow_uuid',
    name: 'Flow to Start'
  },
  create_contact = false
}: {
  uuid?: string;
  groups?: Group[];
  contacts?: Contact[];
  flow?: Flow;
  create_contact?: boolean;
} = {}): StartSession => ({
  uuid,
  groups,
  contacts,
  flow,
  create_contact,
  type: Types.start_session
});

export const createBroadcastMsgAction = ({
  uuid = utils.createUUID(),
  groups = [
    { uuid: utils.createUUID(), name: 'Cat Fanciers' },
    { uuid: utils.createUUID(), name: 'Cat Facts' }
  ],
  contacts = [
    { uuid: utils.createUUID(), name: 'Kellan Alexander' },
    { uuid: utils.createUUID(), name: 'Norbert Kwizera' },
    { uuid: utils.createUUID(), name: 'Rowan Seymour' }
  ],
  text = 'Hello World'
}: {
  uuid?: string;
  groups?: Group[];
  contacts?: Contact[];
  text?: string;
} = {}): BroadcastMsg => ({
  uuid,
  groups,
  contacts,
  text,
  type: Types.send_broadcast
});

export const createAddGroupsAction = ({
  uuid = utils.createUUID(),
  groups = groupsResults
}: { uuid?: string; groups?: Group[] } = {}): ChangeGroups => ({
  uuid,
  type: Types.add_contact_groups,
  groups
});

export const createRemoveGroupsAction = ({
  uuid = utils.createUUID(),
  groups = groupsResults
}: { uuid?: string; groups?: Group[] } = {}): RemoveFromGroups => ({
  uuid,
  all_groups: false,
  type: Types.remove_contact_groups,
  groups
});

export const createStartFlowAction = ({
  uuid = utils.createUUID(),
  flow = {
    name: 'Colors',
    uuid: 'd4a3a01c-1dee-4324-b107-4ac7a21d836f'
  }
}: {
  uuid?: string;
  flow?: {
    name: string;
    uuid: string;
  };
} = {}): StartFlow => ({
  type: Types.enter_flow,
  uuid: 'd4a3a01c-1dee-4324-b107-4ac7a21d836f',
  flow: {
    name: utils.capitalize(flow.name.trim()),
    uuid
  }
});

export const createSetContactNameAction = ({
  uuid = utils.createUUID(),
  name = 'Jane Goodall'
}: {
  uuid?: string;
  name?: string;
} = {}): SetContactProperty => ({
  uuid,
  name,
  type: Types.set_contact_name
});

export const createSetContactFieldAction = ({
  uuid = utils.createUUID(),
  field = {
    key: 'age',
    name: 'Age'
  },
  value = '25'
}: {
  uuid?: string;
  field?: Field;
  value?: string;
} = {}): SetContactField => ({
  uuid,
  field,
  value,
  type: Types.set_contact_field
});

export const createSetContactLanguageAction = ({
  uuid = utils.createUUID(),
  language = 'eng'
}: {
  uuid?: string;
  language?: string;
} = {}): SetContactLanguage => ({
  uuid,
  language,
  type: Types.set_contact_language
});

export const createSetContactChannelAction = ({
  uuid = utils.createUUID(),
  channelName = 'Twilio Channel'
}: {
  uuid?: string;
  channelName?: string;
} = {}): SetContactChannel => ({
  uuid,
  channel: {
    uuid,
    name: channelName
  },
  type: Types.set_contact_channel
});

export const createSetRunResultAction = ({
  uuid = utils.createUUID(),
  name = 'Name',
  value = 'Grace',
  category = ''
}: {
  uuid?: string;
  name?: string;
  value?: string;
  category?: string;
} = {}): SetRunResult => ({
  uuid,
  name,
  value,
  category,
  type: Types.set_run_result
});

export const createWebhookNode = (
  action: CallWebhook | CallResthook | OpenTicket | TransferAirtime,
  useCategoryTest: boolean
) => {
  const { categories, exits } = createCategories([
    WebhookExitNames.Success,
    WebhookExitNames.Failure
  ]);

  const cases: Case[] = [
    {
      uuid: utils.createUUID(),
      type: useCategoryTest ? Operators.has_category : Operators.has_only_text,
      arguments: [WebhookExitNames.Success],
      category_uuid: categories[0].uuid
    }
  ];

  let operand = '@results.' + utils.snakify(action.result_name);
  if (!useCategoryTest) {
    operand += '.category';
  }

  return {
    uuid: utils.createUUID(),
    actions: [action],
    router: {
      type: RouterTypes.switch,
      operand: operand,
      cases,
      categories,
      default_category_uuid: categories[categories.length - 1].uuid
    } as SwitchRouter,
    exits
  };
};

export const createWebhookRouterNode = (): FlowNode => {
  const action: CallWebhook = {
    uuid: utils.createUUID(),
    headers: {},
    type: Types.call_webhook,
    url: 'http://www.google.com',
    method: Methods.GET,
    result_name: 'Response'
  };
  return createWebhookNode(action, false);
};

export const createOpenTicketNode = (subject: string, body: string): FlowNode => {
  const action: OpenTicket = {
    uuid: utils.createUUID(),
    type: Types.open_ticket,
    ticketer: {
      name: 'Email (bob@acme.com)',
      uuid: '1165a73a-2ee0-4891-895e-768645194862'
    },
    subject: subject,
    body: body,
    result_name: 'Result'
  };
  return createWebhookNode(action, true);
};

export const getLocalizationFormProps = (
  action: AnyAction,
  lang?: Asset,
  translations?: { [uuid: string]: any }
): LocalizationFormProps => {
  const language = lang || { id: 'eng', name: 'English', type: AssetType.Language };
  return {
    language,
    onClose: jest.fn(),
    updateLocalizations: jest.fn(),
    issues: [],
    helpArticles: {},
    nodeSettings: {
      originalNode: createRenderNode({
        actions: [action],
        uuid: utils.createUUID(),
        router: null,
        exits: [{ uuid: utils.createUUID() }],
        ui: {
          position: { left: 0, top: 0 },
          type: Types.execute_actions
        }
      }),
      originalAction: action,
      localizations: [Localization.translate(action, language, translations)]
    }
  };
};

export const getActionFormProps = (action: AnyAction): ActionFormProps => ({
  assetStore: {
    channels: { items: {}, type: AssetType.Channel },
    fields: { items: {}, type: AssetType.Field },
    languages: { items: {}, type: AssetType.Language },
    labels: { items: {}, type: AssetType.Label },
    results: { items: {}, type: AssetType.Result },
    flows: { items: {}, type: AssetType.Flow },
    recipients: { items: {}, type: AssetType.Contact || AssetType.Group || AssetType.URN }
  },
  helpArticles: {},
  completionSchema: { root: [], types: [] },
  addAsset: jest.fn(),
  updateAction: jest.fn(),
  onClose: jest.fn(),
  onTypeChange: jest.fn(),
  issues: [],
  typeConfig: getTypeConfig(action.type),
  nodeSettings: {
    originalNode: createRenderNode({
      actions: [action],
      uuid: utils.createUUID(),
      router: null,
      exits: [{ uuid: utils.createUUID() }],
      ui: {
        position: { left: 0, top: 0 },
        type: Types.execute_actions
      }
    }),
    originalAction: action
  }
});

export const getRouterFormProps = (renderNode: RenderNode): RouterFormProps => ({
  helpArticles: {},
  updateRouter: jest.fn(),
  onClose: jest.fn(),
  onTypeChange: jest.fn(),
  issues: [],
  typeConfig: determineTypeConfig({ originalNode: renderNode }),
  assetStore: EMPTY_TEST_ASSETS,
  nodeSettings: {
    originalNode: renderNode,
    originalAction: null
  }
});

// tslint:disable-next-line:variable-name
export const createCase = ({
  uuid,
  type,
  category_uuid,
  args = []
}: {
  uuid: string;
  type: Operators;
  category_uuid: string;
  args?: string[];
}) => ({
  uuid,
  type,
  category_uuid,
  arguments: args
});

export const createExit = ({
  uuid = utils.createUUID(),
  destination_uuid = null
}: {
  uuid?: string;
  destination_uuid?: string;
} = {}) => ({
  uuid,
  destination_uuid
});

// tslint:disable-next-line:variable-name
export const createRouter = (result_name?: string): Router => ({
  categories: [],
  type: RouterTypes.switch,
  ...(result_name ? { result_name } : {})
});

export const createMatchCase = (match: string): CaseProps => {
  return {
    uuid: utils.createUUID(),
    categoryName: match,
    valid: true,
    kase: {
      uuid: utils.createUUID(),
      arguments: [match.toLowerCase()],
      type: Operators.has_any_word,
      category_uuid: null
    }
  };
};

export const createEmptyNode = (): FlowNode => {
  return {
    uuid: utils.createUUID(),
    actions: [],
    exits: []
  };
};

export const createCases = (categories: string[]): CaseProps[] => {
  const cases: CaseProps[] = [];
  categories.forEach((category: string) => {
    cases.push(createMatchCase(category));
  });
  return cases;
};

export const createRoutes = (categories: string[], hasTimeout: boolean = false): ResolvedRoutes => {
  const cases: CaseProps[] = [];
  categories.forEach((category: string) => {
    cases.push(createMatchCase(category));
  });

  return resolveRoutes(cases, hasTimeout, null);
};

export const createWaitRouter = (hintType: HintTypes, resultName: string = 'Result Name') => {
  const originalNode = createMatchRouter([]);
  originalNode.node.router.wait = { type: WaitTypes.msg, hint: { type: hintType } };
  originalNode.node.router.result_name = resultName;
  return originalNode;
};

export const createMatchRouter = (matches: string[], hasTimeout: boolean = false): RenderNode => {
  const { exits, categories, cases, timeoutCategory } = createRoutes(matches, hasTimeout);

  const wait: Wait = hasTimeout
    ? {
        type: WaitTypes.msg,
        timeout: { seconds: 60, category_uuid: timeoutCategory }
      }
    : { type: WaitTypes.msg };

  return createRenderNode({
    actions: [],
    exits,
    router: {
      type: RouterTypes.switch,
      operand: DEFAULT_OPERAND,
      categories,
      cases,
      wait,
      default_category_uuid: categories[categories.length - 1].uuid,
      result_name: ''
    } as SwitchRouter,
    ui: {
      type: Types.wait_for_response,
      position: { left: 0, top: 0 }
    }
  });
};

export const createSchemeRouter = (schemes: Scheme[]): RenderNode => {
  const matchRouter = createMatchRouter(schemes.map((scheme: Scheme) => scheme.scheme));

  const router = matchRouter.node.router as SwitchRouter;

  // switch our operators to be consistent with scheme routers
  // set our operand and type for scheme matching
  delete router['wait'];
  router.cases.forEach((kase: Case) => {
    kase.type = Operators.has_only_phrase;
    const category = router.categories.find(
      (category: Category) => category.uuid === kase.category_uuid
    );
    category.name = schemes.find((scheme: Scheme) => scheme.scheme === kase.arguments[0]).name;
  });

  router.operand = SCHEMES_OPERAND;
  matchRouter.ui.type = Types.split_by_scheme;

  return matchRouter;
};

export const createSwitchRouter = ({
  cases,
  categories = [],
  operand = '@input',
  wait = null,
  default_category_uuid = null
}: {
  cases: Case[];
  categories: Category[];
  operand?: string;
  wait?: Wait;
  // tslint:disable-next-line:variable-name
  default_category_uuid?: string;
}) => ({
  ...createRouter(),
  cases,
  categories,
  operand,
  wait,
  default_category_uuid
});

export const createRenderNode = ({
  actions,
  exits,
  uuid = utils.createUUID(),
  router = null,
  ui = {
    position: { left: 0, top: 0 },
    type: Types.split_by_expression
  }
}: {
  actions: AnyAction[];
  exits: Exit[];
  uuid?: string;
  router?: Router | SwitchRouter;
  ui?: UINode;
}): RenderNode => {
  const renderNode: RenderNode = {
    node: {
      actions,
      exits,
      uuid,
      ...(router ? { router } : ({} as any))
    },
    ui,
    inboundConnections: {}
  };
  return renderNode;
};

export const createFlowNode = ({
  actions,
  exits,
  uuid = utils.createUUID(),
  router = null,
  wait = null
}: {
  actions: AnyAction[];
  exits: Exit[];
  uuid?: string;
  router?: Router | SwitchRouter;
  wait?: Wait;
}): FlowNode => ({
  actions,
  exits,
  uuid,
  ...(router ? { router } : ({} as any)),
  ...(wait ? { wait } : ({} as any))
});

export const createCategories = (names: string[]): { categories: Category[]; exits: Exit[] } => {
  const exits = names.map((cat: string) => {
    return {
      uuid: utils.createUUID(),
      destination_uuid: null
    };
  });

  const categories = exits.map((ex: Exit, index: number) => {
    return {
      name: names[index],
      uuid: utils.createUUID(),
      exit_uuid: ex.uuid
    };
  });

  return { exits, categories };
};

export const createRandomNode = (buckets: number): RenderNode => {
  const { categories, exits } = createCategories(
    utils.range(0, buckets).map((bucketIdx: number) => `Bucket ${bucketIdx + 1}`)
  );
  return createRenderNode({
    actions: [],
    exits,
    uuid: utils.createUUID(),
    router: {
      type: RouterTypes.random,
      categories
    },
    ui: { position: { left: 0, top: 0 }, type: Types.split_by_random }
  });
};

export const createSubflowNode = (
  startFlowAction: StartFlow,
  uuid: string = utils.createUUID()
): RenderNode => {
  const { categories, exits } = createCategories([
    StartFlowExitNames.Complete,
    StartFlowExitNames.Expired
  ]);

  return createRenderNode({
    actions: [startFlowAction],
    exits,
    uuid,
    router: createSwitchRouter({
      categories,
      cases: [
        createCase({
          uuid: utils.createUUID(),
          type: Operators.has_run_status,
          category_uuid: categories[0].uuid,
          args: [StartFlowArgs.Complete]
        }),
        createCase({
          uuid: utils.createUUID(),
          type: Operators.has_run_status,
          category_uuid: categories[1].uuid,
          args: [StartFlowArgs.Expired]
        })
      ],
      operand: '@child',
      default_category_uuid: null
    }),
    ui: { position: { left: 0, top: 0 }, type: Types.split_by_subflow }
  });
};

export const createAirtimeTransferNode = (transferAirtimeAction: TransferAirtime): RenderNode => {
  return {
    node: createWebhookNode(transferAirtimeAction, true),
    ui: { position: { left: 0, top: 0 }, type: Types.split_by_airtime },
    inboundConnections: {}
  };
};

export const createResthookNode = (callResthookAction: CallResthook): RenderNode => {
  return {
    node: createWebhookNode(callResthookAction, false),
    ui: { position: { left: 0, top: 0 }, type: Types.split_by_resthook },
    inboundConnections: {}
  };
};

export const createGroupsRouterNode = (
  groups: Group[] = groupsResults,
  uuid: string = utils.createUUID()
): RenderNode => {
  const exits = groups.map((group, idx) =>
    createExit({
      uuid: utils.createUUID(),
      destination_uuid: null
    })
  );

  exits.push({ uuid: utils.createUUID(), destination_uuid: null });

  const categories = groups.map((group, idx) => {
    return {
      name: group.name,
      uuid: utils.createUUID(),
      exit_uuid: exits[0].uuid
    };
  });

  categories.push({
    name: DefaultExitNames.Other,
    uuid: utils.createUUID(),
    exit_uuid: exits[exits.length - 1].uuid
  });

  const cases = groups.map((group, idx) =>
    createCase({
      uuid: utils.createUUID(),
      type: Operators.has_group,
      category_uuid: categories[idx].uuid,
      args: [group.uuid]
    })
  );

  return createRenderNode({
    actions: [],
    exits,
    uuid,
    router: createSwitchRouter({
      categories,
      cases,
      operand: GROUPS_OPERAND,
      default_category_uuid: categories[categories.length - 1].uuid
    }),
    ui: {
      type: Types.split_by_groups,
      position: { left: 0, top: 0 }
    }
  });
};

export const getGroupOptions = (groups: Group[] = groupsResults) =>
  groups.map(({ name, uuid }) => ({
    name,
    id: uuid
  }));

export const getGroups = (sliceAt: number, groups: Group[] = groupsResults) =>
  groups
    .map(({ name, uuid }) => ({
      name,
      id: uuid
    }))
    .slice(sliceAt);

export const createAddLabelsAction = (labels: Label[]) => ({
  type: Types.add_input_labels,
  uuid: 'aa15ef19-da81-43d0-b6e5-84b47216aeb8',
  labels
});

export const English = { name: 'English', id: 'eng', type: AssetType.Language };

export const Spanish = { name: 'Spanish', id: 'spa', type: AssetType.Language };

export const SubscribersGroup = {
  name: 'Subscriber',
  id: '68223118-109f-442a-aed3-7bb3e1eab687',
  type: AssetType.Group
};

export const ColorFlowAsset = {
  name: 'Favorite Color',
  uuid: '9a93ede6-078f-44c9-ad0a-133793be5d56',
  content: {
    parent_refs: ['colors']
  }
};

export const ResthookAsset = {
  id: 'new-resthook',
  name: 'new-resthook',
  type: AssetType.Resthook
};

export const FeedbackLabel = {
  name: 'Feedback',
  id: 'feedback_label',
  type: AssetType.Label
};

export const languages: Assets = {
  items: assetListToMap(languagesResults.results.map((language: any) => languageToAsset(language))),
  type: AssetType.Language
};
