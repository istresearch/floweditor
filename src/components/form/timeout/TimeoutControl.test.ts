import TimeoutControl, {
  TIMEOUT_OPTIONS,
  TimeoutControlProps
} from 'components/form/timeout/TimeoutControl';
import { composeComponentTestUtils, setMock } from 'testUtils';

const { setup } = composeComponentTestUtils<TimeoutControlProps>(TimeoutControl, {
  timeout: 0,
  onChanged: jest.fn()
});

describe(TimeoutControl.name, () => {
  it('renders', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('updates', () => {
    const { instance, props } = setup(true, { onChanged: setMock() });
    instance.handleChecked();
    expect(props.onChanged).toMatchCallSnapshot('check');

    instance.handleTimeoutChanged(TIMEOUT_OPTIONS[0]);
    expect(props.onChanged).toMatchCallSnapshot('update');
  });

  it('handles initial values', () => {
    const { wrapper, instance } = setup(true, {
      timeout: { $set: parseInt(TIMEOUT_OPTIONS[0].value) },
      onChanged: setMock()
    });

    instance.handleChecked();
    expect(wrapper).toMatchSnapshot();
  });
});
