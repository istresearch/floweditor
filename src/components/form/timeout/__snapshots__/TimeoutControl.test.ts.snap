// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeoutControl handles initial values 1`] = `
<div
  className="timeout_control_container"
>
  <div
    className="left_section"
  >
    <CheckboxElement
      checkboxClassName="checkbox"
      checked={true}
      description="Continue when there is no response for"
      name="Timeout"
      onChange={[Function]}
    />
  </div>
  <div
    className="drop_down"
  >
    <TembaSelect
      name="Timeout"
      onChange={[Function]}
      options={
        Array [
          Object {
            "name": "1 minute",
            "value": "60",
          },
          Object {
            "name": "2 minutes",
            "value": "120",
          },
          Object {
            "name": "3 minutes",
            "value": "180",
          },
          Object {
            "name": "4 minutes",
            "value": "240",
          },
          Object {
            "name": "5 minutes",
            "value": "300",
          },
          Object {
            "name": "10 minutes",
            "value": "600",
          },
          Object {
            "name": "15 minutes",
            "value": "900",
          },
          Object {
            "name": "1 hour",
            "value": "3600",
          },
          Object {
            "name": "2 hours",
            "value": "7200",
          },
          Object {
            "name": "3 hours",
            "value": "10800",
          },
          Object {
            "name": "6 hours",
            "value": "21600",
          },
          Object {
            "name": "12 hours",
            "value": "43200",
          },
          Object {
            "name": "18 hours",
            "value": "64800",
          },
          Object {
            "name": "1 day",
            "value": "86400",
          },
          Object {
            "name": "2 days",
            "value": "172800",
          },
          Object {
            "name": "3 days",
            "value": "259200",
          },
          Object {
            "name": "1 week",
            "value": "604800",
          },
        ]
      }
      style="small"
      value={
        Object {
          "name": "1 minute",
          "value": "60",
        }
      }
    />
  </div>
</div>
`;

exports[`TimeoutControl renders 1`] = `
<div
  className="timeout_control_container"
>
  <div
    className="left_section"
  >
    <CheckboxElement
      checkboxClassName="checkbox"
      checked={false}
      description="Continue when there is no response..."
      name="Timeout"
      onChange={[Function]}
    />
  </div>
</div>
`;

exports[`TimeoutControl updates: check 1`] = `
Array [
  300,
]
`;

exports[`TimeoutControl updates: update 1`] = `
Array [
  300,
]
`;
